nntrainer/models/neuralnet.cpp

```diff
@@ -632,25 +632,26 @@ int NeuralNetwork::train_run() {
     return ML_ERROR_INVALID_PARAMETER;
   }
 
-  for (epoch_idx = epoch_idx + 1; epoch_idx <= epochs; ++epoch_idx) {
-    training.loss = 0.0f;
-
+  /**
+   * @brief run a single epoch with given callback, @a auto is used instead of
+   * std::function for performance measure
+   * @param buffer buffer to run
+   * @param shuffle whether to shuffle or not
+   * @param on_iteration_fetch function that will recieve reference to stat,
+   * buffer which will be called every time data is fetched and set
+   * @param on_epoch_end function that will recieve reference to stat,
+   * buffer which will be called on the epoch end
+   */
+  auto run_epoch = [this, &in, &label, &in_dims, &label_dims](
```

Comment:
💯

Reply:
you can consider making it a private function itself than a lambda as there will be eventually many hooks supported for this.

Reply:
I agree, But i'll delay for now :)

Reply:
IMO, the function will most likely be a static function which will additionally take `NetworkGraph` object, and return stats as now, so can remain in .cpp.

Reply:
If it remains in .cpp with static, I guess this can stay here until we have the need.